near/omni-prover/evm-prover-proxy/src/lib.rs

```diff
@@ -0,0 +1,103 @@
+use near_sdk::borsh::{self, BorshDeserialize, BorshSerialize};
+use near_sdk::serde::{Deserialize, Serialize};
+use near_sdk::{
+    env, ext_contract, near_bindgen, AccountId, Gas, PanicOnDefault, Promise, PromiseError,
+};
+use omni_types::evm_events::parse_evm_event;
+use omni_types::prover_types::{ProofKind, ProofResult};
+
+/// Gas to call verify_log_entry on prover.
+pub const VERIFY_LOG_ENTRY_GAS: Gas = Gas::from_tgas(50);
+
+#[ext_contract(ext_prover)]
+pub trait Prover {
+    #[result_serializer(borsh)]
+    fn verify_log_entry(
+        &self,
+        #[serializer(borsh)] log_index: u64,
+        #[serializer(borsh)] log_entry_data: Vec<u8>,
+        #[serializer(borsh)] receipt_index: u64,
+        #[serializer(borsh)] receipt_data: Vec<u8>,
+        #[serializer(borsh)] header_data: Vec<u8>,
+        #[serializer(borsh)] proof: Vec<Vec<u8>>,
+        #[serializer(borsh)] skip_bridge_call: bool,
+    ) -> bool;
+}
+
+#[derive(Default, BorshDeserialize, BorshSerialize, Clone, Serialize, Deserialize)]
+#[serde(crate = "near_sdk::serde")]
+pub struct Proof {
+    pub log_index: u64,
+    pub log_entry_data: Vec<u8>,
+    pub receipt_index: u64,
+    pub receipt_data: Vec<u8>,
+    pub header_data: Vec<u8>,
+    pub proof: Vec<Vec<u8>>,
+}
+
+#[near_bindgen]
+#[derive(BorshDeserialize, BorshSerialize, PanicOnDefault)]
+pub struct RainbowOmniProverProxy {
+    pub prover_account: AccountId,
+}
+
+#[near_bindgen]
+impl RainbowOmniProverProxy {
+    #[init]
+    #[private]
+    #[must_use]
+    pub fn init(prover_account: AccountId) -> Self {
+        Self { prover_account }
+    }
+
+    pub fn verify_proof(
+        &self,
+        #[serializer(borsh)] kind: ProofKind,
+        #[serializer(borsh)] msg: Vec<u8>,
+    ) -> Promise {
+        let proof =
+            Proof::try_from_slice(&msg).unwrap_or_else(|_| env::panic_str("ErrorOnProofParsing"));
+
+        ext_prover::ext(self.prover_account.clone())
+            .with_static_gas(VERIFY_LOG_ENTRY_GAS)
+            .verify_log_entry(
+                proof.log_index,
+                proof.log_entry_data.clone(),
+                proof.receipt_index,
+                proof.receipt_data,
+                proof.header_data,
+                proof.proof,
+                false, // Do not skip bridge call. This is only used for development and diagnostics.
+            )
+            .then(
+                Self::ext(env::current_account_id())
+                    .with_static_gas(VERIFY_LOG_ENTRY_GAS)
+                    .verify_log_entry_callback(kind, proof.log_entry_data),
```

Comment:
`kind` is not validated in any way. Is it possible to pass kind different from the actual proof and force a different type of result in the callback

Reply:
No, it isn't possible because the `parse_evm_event` will fail in this case

Reply:
What if two events have the same structure (parameter types and order)?

Reply:
@kiseln The first topic is the signature of the name of the event that occurred.
So I suppose that the decode logs verify the Events topics types because we pass `validate=true` https://docs.rs/alloy-sol-types/latest/alloy_sol_types/trait.SolEvent.html#method.decode_log

Reply:
This is not clear to me looking at the library code. Perhaps a test would be useful here

Reply:
@kiseln yeh, it is true, I've added a verification here dd7d84ccbe96e4ed3628d286cf55dd392b5f049c